---
title: "GLM_Tutorial"
author: "Rachael Kane"
date: "3/20/2020"
output: html_document
vignette: >
  %\VignetteIndexEntry{GLM_Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


GLM stands for general linear model. GLM's are a framework used to compare several variables. GLM's have increased power when compared to t-test association evaluation and are used for a varitey of analyses. In this tutorial you will learn how to apply a GLM to test for significant associations between quantitative trait nucleotides (QTNs) and observed phenotypes. This is called a genome wide association study (GWAS) were the desired outcome is identification of QTNs with significant associations with traits of interest.  


Steps in tutorial
1. Download and manipulate demo data
2. Obtain principal components, sample QTNs and simulate phenotypes utilizing G2P function
3. Use GLM function to perform GWAS by GLM and make a manhattan plot to visualize QTN positions 
4. Make a QQ plot to determine whether pvalues are inflated for demo data


1. Download and manipulte demo data

Demo data is sourced from zzlab.net. The demo data is manipulated to obtain the appropriate input format. Then the demo data is filtered by a minor allele frequency of 0.05 which is a standard within scienctific literature. 

```{r setup, include=FALSE}

#to download demo data and have R read as a table
myGD=read.table(file="http://zzlab.net/GAPIT/data/mdp_numeric.txt",head=T)
myGM=read.table(file="http://zzlab.net/GAPIT/data/mdp_SNP_information.txt",head=T)

#myGD should consist of 281 observations of 3094 variables. These represent 281 individuals with 3093 genetic markers.
#myGM should consist of 3093 observations of 3 variables. These represent 3093 genetic markers and columns with SNP label, chromosome #, and base pair position.
#phenotypes should consist of 281 observations of 2 variables. These represent the 281 individuals taxa ID and their observaed phenotype score.
#covariates should consist of

#to make the myGD and myGM data the appropriate input format --> remove taxa ID column from myGD and transpose myGM
myGD_temp<-myGD[,-1]
myGD_transposed<-t(myGD_temp)

#now myGD should consist of 281 observations of 3093 variables which now matches the genetic markers listed in myGM

#add column for minor allele frequency
maf<-vector(length=nrow(myGD_transposed))
myGD_transposed<-cbind(myGD_transposed, maf)

#determine minor allele frequency for each genetic marker
for (i in 1:nrow(myGD_transposed)){
  freq1<-(((2*(sum((myGD_transposed[i,1:(ncol(myGD_transposed)-1)])==0)))+(1*(sum((myGD_transposed[i,1:(ncol(myGD_transposed)-1)])==1))))/(2*((ncol(myGD_transposed)))))
  freq2<-(((2*(sum((myGD_transposed[i,1:(ncol(myGD_transposed)-1)])==2)))+(1*(sum((myGD_transposed[i,1:(ncol(myGD_transposed)-1)])==1))))/(2*((ncol(myGD_transposed)))))
  myGD_transposed[i,ncol(myGD_transposed)]<-min(freq1,freq2)
}

#temporarily bind myGM and myGD_transposed together --> to apply MAF filter to both at same time
GM_GD_full<-cbind(myGM, myGD_transposed)

#remove markers with minor allele frequency =< 0.05
GM_GD_full<-subset(GM_GD_full, GM_GD_full[,ncol(GM_GD_full)]>0.05)

#remove column containing minor allele frequencies
GM_GD_full2<-GM_GD_full[,1:(ncol(GM_GD_full)-1)]

#creates new marker map
myGM_new<-GM_GD_full2[,1:3]

#creates new genotype data matrix
myGD_transposed_new<-GM_GD_full2[,4:ncol(GM_GD_full2)]
myGD_untransposed<-t(myGD_transposed_new)
X=myGD_untransposed

knitr::opts_chunk$set(echo = TRUE)
```

2. Obtain principal components, sample QTNs and simulate phenotypes utilizing G2P function

```{r, echo=FALSE}
cofactor.pca.cor<-function(U, G){
  #Carries out principal components analysis
  pca.obj<-prcomp(G)
  #Isolates the principal component scores matrix (rows as individuals, columns as principal components)
  pca<-pca.obj$x

  #If user-specified cofactors (U) are not specified, the function returns the principal component scores matrix
  if(missing(U)){
    gwas.covariates<-pca

    #Combines the original principal components scores, the final set of covariates (U + retained principal components)
    list_cov<-list(cov=gwas.covariates)

    #Output$cov is a covariate matrix for use as the argument "C" in the GWASbyGLM function
    return(list_cov)

    #If user-specified cofactors (U) are specified, the function tests for correlations between cofactors in U and principal components
  }else{
    #Borrows the matrix correlation function from the R package "psych"
    pca.c.corr.test<-corr.test(x=U[,1:ncol(U)], y=pca[,1:ncol(pca)], adjust="none")

    #Identifies pairs of U cofactors and principal components that are significantly correlated, with a Bonferroni correction for multiple testing
    #Columns in sig.pca.c.corr are principal components, rows are U cofactors
    #The sig.pca.c.corr matrix cells contain values of 1 and 0, indicating significant correlation or lack of correlation, respectively, between the principal component and U cofactor
    sig.pca.c.corr<-pca.c.corr.test$p<(0.05/(ncol(U)*ncol(pca)))

    #Creates empy matrix, to which the retained principal components and individual scores will be attached
    filtered.pca.temp<-matrix(ncol=1,nrow=nrow(pca))

    #Creates empty dataframe, to be filled with lines indicating which principal components are removed
    removal.report.temp<-matrix(ncol=1, nrow=1)

    #When a principal component is correlated with any of the U cofactors, it is removed
    for (i in ncol(sig.pca.c.corr)){
      #Columns in sig.pca.c.corr are principal components, rows are U cofactors
      #If a principal component is uncorrelated with all U cofactors, the sum down the column equals 0
      if ((sum(sig.pca.c.corr[,i]))==0){
        filtered.pca.temp<-cbind(filtered.pca.temp, pca[,i])
      }else{
        report<-paste("Removed principal component", i)
        removal.report.temp<-rbind(removal.report.temp, report)
      }
    }

    #Creates matrix consisting of principal components and individual scores for the uncorrelated principal components
    filtered.pca<-data.matrix(filtered.pca.temp[,2:ncol(filtered.pca.temp)])

    #Binds the matrix of U cofactors with the matrix of retained principal components
    gwas.covariates<-cbind(U,filtered.pca)

    #Creates the final report of which principal components were removed
    removal.report<-removal.report.temp[2:nrow(removal.report.temp),1]

    #Combines the original principal components scores, the final set of covariates (U + retained principal components, and the removal report)
    list_origpca_retainedcov_removed<-list(orig_pc=pca, cov=gwas.covariates, removed=removal.report)

    #Function returns this set of outputs when U is specified
    #Output$cov is a covariate matrix for use as the argument "C" in the GWASbyGLM function
    return(list_origpca_retainedcov_removed)
  }
}


```


```{r}

source("http://zzlab.net/StaGen/2020/R/G2P.R")

#obtain principal components
#the argument U is unspecified, so the cofactor.pca.cor function will essentially carry out PCA and return principal components and individual scores

principal_components<-cofactor.pca.cor(G=X)

#screeplot to determine how many principal components to retain (here we should retain about 3)
pca_X<-prcomp(X)
pca_X_var<-(pca_X$sdev^2)
pca_X_pvar<-(pca_X_var/sum(pca_X_var))
plot(pca_X_pvar,xlab="Principal component", ylab="Proportion of variance explained", ylim=c(0,0.07), type='b')

#simulate phenotypes
set.seed(99164)
mySim=G2P(X, h2=.75,alpha=1,NQTN=10,distribution="normal")

##### plot QTN position
plot(myGM_new[,c(2,3)], xlab="Chromosome", ylab="Position")
lines(myGM_new[mySim$QTN.position, c(2,3)], type="p", col="red")
points(myGM_new[mySim$QTN.position, c(2,3)], type="p", col="blue", cex=5)

```

3. Use GLM function to perform GWAS by GLM and make a manhattan plot to visualize significant SNPs


```{r}

#carry out GLM
#set values for arguments
y=mySim$y
G=X
C=principal_components$cov
NC=3


glm_test<-GWASbyGLM(y=y, G=G, C=C, NC=NC)

#Manhattan plot
color.vector <- rep(c("deepskyblue","orange","forestgreen","indianred3"),9)
m=nrow(myGM_new)
plot(seq(1:m),-log10(glm_test), col=color.vector[myGM_new[,2]], ylab="-log(P)", xlab="Marker position", sub="Vertical lines indicate QTN position", main="GWAS by GLM p-value vs position")
abline(v=mySim$QTN.position, lty = 2, lwd=1.5, col = "black")
legend(4500, 9, legend=c("QTN"),
       col=c("black"), lty=c(2), cex=0.8)

#identify QTNs among top-10 SNPs
#orders markers by p-value from smallest to largest
index_glm_test=order(glm_test)
#identifies the 10 markers with the lowest p-values
top10_glm_test=index_glm_test[1:10]
#identifies QTNs among the top 10 markers
detected_glm_test=intersect(top10_glm_test, mySim$QTN.position)
#returns the number of true positives among the top 10 markers
length(detected_glm_test)


```

4. Make a QQ plot to determine whether pvalues are inflated

Manhattan plots are a type of scatter plot, usually used to visualize data with a large number of data-points. In this tutorial it is being used to display statistically significant SNPs found in our GWASbyGLM analysis. The SNPs found above the threshold are true QTNs.


QQ plots are used to determine if the pvalues you see in your data are inflated based on the expected pvalues.The red line on the graph represents the expected pvalues while the black circles represents the pvalues associated with your data. 

```{r}

## QQplot
p.obs=glm_test
m2=length(p.obs)
# m2 = markers in the last chrom
p.uni=runif(m2,0,1)
order.obs=order(p.obs)
order.uni=order(p.uni)

plot(-log10(p.uni[order.uni]),-log10(p.obs[order.obs]), main = "QQPlot")
abline(a = 0, b = 1, col = "red")


'''
