---
title: "GLM_Tutorial"
author: "Rachael Kane"
date: "3/20/2020"
output: html_document
vignette: >
  %\VignetteIndexEntry{GLM_Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

GLM stands for general linear model. GLM's are a framework used to compare several variables. GLM's have increased power when compared to t-test association evaluation and are used for a varitey of metrics. In this tutorial you will learn how to apply a GLM to test for significant associations between quantitative trait nucleotides and observed phenotypes.

The general equation for a GLM is below.

Y = SNP + Q + e

Y: observed phenotype
SNP: fixed effect (true QTNs and SNPs)
Q: fixed effect (population structure)
e: residual

Steps in tutorial
1. Download and manipulate demo data
2. Sample QTNs and simulate phenotypes utilizing G2P function
3. Use GLM function to perform GWAS by GLM
4. Make manhattan plot and QQ plot and determine true QTNs and whether pvalues are inflated


1. Download and manipulte demo data

```{r setup, include=FALSE}
#set your working directory to where ever you wish to have the demo data download

#to download demo data and have R read as a table
myGD=read.table(file="http://zzlab.net/GAPIT/data/mdp_numeric.txt",head=T)
myGM=read.table(file="http://zzlab.net/GAPIT/data/mdp_SNP_information.txt",head=T)
covariates=read.table(file = "http://zzlab.net/GAPIT/data/CROP545_Covariates.txt", head=T)

#myGD should consist of 281 observations of 3094 variables. These represent 281 individuals with 3093 genetic markers.
#myGM should consist of 3093 observations of 3 variables. These represent 3093 genetic markers and columns with SNP label, chromosome #, and base pair position.
#phenotypes should consist of 281 observations of 2 variables. These represent the 281 individuals taxa ID and their observaed phenotype score.
#covariates should consist of

#to make the myGD data the appropriate input format for the G2P function--> remove taxa ID column
myGD <- cbind(myGD[,2:3094])
#now myGD should consist of 281 observations of 3093 variables which now matches the genetic markers listed in myGM

knitr::opts_chunk$set(echo = TRUE)
```

2. Sample QTNs and simulate phenotypes utilizing G2P function

```{r cars}

#define function 
G2P=function(X,h2,alpha,NQTN,distribution){
n=nrow(X)
m=ncol(X)
#Sampling QTN
QTN.position=sample(m,NQTN,replace=F)
SNPQ=as.matrix(X[,QTN.position])
QTN.position
#QTN effects
if(distribution=="norm")
	{addeffect=rnorm(NQTN,0,1)
	}else
	{addeffect=alpha^(1:NQTN)}
#Simulate phenotype
effect=SNPQ%*%addeffect
effectvar=var(effect)
residualvar=(effectvar-h2*effectvar)/h2
residual=rnorm(n,0,sqrt(residualvar))
y=effect+residual
return(list(addeffect = addeffect, y=y, add = effect, residual = residual, QTN.position=QTN.position, SNPQ=SNPQ))
}


#simulate a phenotype
NQTN = 10 # number of QTNs to select from your dataset 
h2 = 0.75 #heritability --> can be changed to different values
alpha = 0.6 
distribution = "norm" 
set.seed(1337) # Fix randomization --> can be set to different seed values
sim_pheno <- G2P(myGD, h2, alpha, NQTN, distribution)

```


3. Use GLM function to perform GWAS by GLM

```{r pressure, echo=FALSE}

Y = sim_pheno$y  #matrix of phenotypes produced by G2P function
G = myGD   #original genotype data
C =        #covariate data imported from zzlab.net
PC = 3     #number of principle components you wish to retain after filtering for correlation with covariates 
plot(pressure)
```

4. Make manhattan plot and QQ plot and determine true QTNs and whether pvalues are inflated

Manhattan plots are a type of scatter plot, usually used to visualize data with a large number of data-points. In this tutorial it is being used to display statistically significant SNPs found in our GWASbyGLM analysis. The SNPs found above the threshold are true QTNs.


QQ plots are used to determine if the pvalues you see in your data are inflated based on the expected pvalues.The red line on the graph represents the expected pvalues while the black circles represents the pvalues associated with your data. 

```{r pressure, echo=FALSE}

#this makes a vector with a different color for each chromosome for the manhattan plot
color.vector <- rep(c("deepskyblue","orange","forestgreen"),3)
m=ncol(myGD)
plot(seq(1:m),-log10(GWASbyGLM), col=color.vector[myGM[,2]], main = "manhattan plot")
abline(v=sim_pheno_GLM$QTN.position, lty = 2, lwd=1.5, col = "black")

## QQplot
p.obs=GWASbyGLM
m2=length(p.obs)
# m2 = markers in the last chrom
p.uni=runif(m2,0,1)
order.obs=order(p.obs)
order.uni=order(p.uni)

plot(-log10(p.uni[order.uni]),-log10(p.obs[order.obs]), main = "QQPlot")
abline(a = 0, b = 1, col = "red")

plot(pressure)
```
